sls-versions/src/main/java/com/palantir/sls/versions/SlsVersionMatcherParser.java

```diff
@@ -0,0 +1,146 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.sls.versions;
+
+import java.util.Optional;
+import java.util.OptionalInt;
+
+/**
+ * A hand-rolled implementation of {@code RegexSlsVersionMatcherParser}.
+ *
+ * Uses two-element int[] arrays as the 'result' type to represent a successful/failed parse, where result[0] always
+ * contains the index into the string we're looking at, and result[1] contains the value returned by the parser
+ * combinator.
+ */
+final class SlsVersionMatcherParser {
+
+    private static final int MAGIC_X_NUMBER = -1;
+
+    public static Optional<SlsVersionMatcher> safeValueOf(String string) {
+        OptionalInt major = OptionalInt.empty();
+        OptionalInt minor = OptionalInt.empty();
+        OptionalInt patch = OptionalInt.empty();
+
+        // major
+        int[] majorResult = numberOrX(string, 0);
+        if (majorResult[1] == Integer.MIN_VALUE) {
+            return Optional.empty(); // reject
+        }
+        if (majorResult[1] != MAGIC_X_NUMBER) {
+            major = OptionalInt.of(majorResult[1]);
+        }
+
+        // dot
+        int[] dot = literalDot(string, majorResult[0]);
+        if (dot[1] == Integer.MIN_VALUE) {
+            return Optional.empty();
+        }
+
+        // minor
+        int[] minorResult = numberOrX(string, dot[0]);
+        if (minorResult[1] == Integer.MIN_VALUE) {
+            return Optional.empty(); // reject
+        }
+        if (minorResult[1] != MAGIC_X_NUMBER) {
+            minor = OptionalInt.of(minorResult[1]);
+        }
+
+        // dot
+        int[] secondDot = literalDot(string, minorResult[0]);
+        if (secondDot[1] == Integer.MIN_VALUE) {
+            return Optional.empty();
+        }
+
+        // patch
+        int[] patchResult = numberOrX(string, secondDot[0]);
+        if (patchResult[1] == Integer.MIN_VALUE) {
+            return Optional.empty(); // reject
+        }
+        if (patchResult[1] != MAGIC_X_NUMBER) {
+            patch = OptionalInt.of(patchResult[1]);
+        }
+
+        if (patchResult[0] < string.length()) {
+            return Optional.empty(); // reject due to trailing stuff
+        }
+
+        return SlsVersionMatcher.maybeCreate(string, major, minor, patch);
+    }
+
+    // "x" is signified by the magic negative number -1, which is distinct from Integer.MIN_VALUE which is a failure
+    private static int[] numberOrX(String string, int startIndex) {
+        int[] result = literalX(string, startIndex);
+        if (result[1] != Integer.MIN_VALUE) {
+            result[1] = MAGIC_X_NUMBER;
+            return result;
+        }
+
+        result = number(string, startIndex);
+        if (result[1] != Integer.MIN_VALUE) {
+            return result;
+        }
+
+        return new int[] {startIndex, Integer.MIN_VALUE};
```

Comment:
I think if you use a type instead of an array it might be easier to read and also less memory, as arrays are objects in java but also contain an extra length field that wouldn't be needed with a type.

Reply:
Ah, apparently HotSpot treats arrays specially and stores the length in object header, so I don't think it would be more memory efficient.